### DISCUSSION TITLE
Use multiple languages for API routes in Next.js / multi-language runtime

### DISCUSSION THREAD

**Discussion by eric-burel** [vercel/vercel#6197] [Ideas] [answered]
Hi, I've been running a few experiments with Next + Python API routes, documented there: https://github.com/lbke/snext.
From the project Readme, here is the behaviour I observe using `vercel dev`:

Vercel CLI command `vercel dev` is able to simulate a serverless environment locally, so you can test your API routes.

- If you put all routes in `./api/`, `vercel dev` will make Python API routes available + Next pages
- If you put all API routes in `./pages/api`, `vercel dev` will make Node.js API routes available + Next pages

**This means you have to pick either Python or Node for your API routes, you can't mix both as far as I can tell.**

Even if it might be an edge case, it would be interesting to be able to support multiple language, because you can end up with different type of API routes in the same project, say:

- I create a dashboard app using Next.js
- I use pandas to preprocess some data for display
- I use Stripe Node API to process some payments

Is that realistic? Could Vercel and `vercel dev` support multiple runtimes for the same app, on a per-route basis? It could take the form of a "Runtime" merger for instance, whose role would be limited to trigger other language-specific runtimes + redirect the request accordingly.

**Comment by eric-burel** [comment, marked as answer]
Ok I seem to have mistested, it already works perfectly when putting the `api/` folder at the root and using `vercel dev`, I can mix Python and JS

**Reply by wakywayne** [reply]
So you couldn't do this by using the regular pages/api folder you had to do something different? I am trying to figure out how to write Golang code with nextJS

**Reply by amyegan** [reply]
Hi @wakywayne! Serverless Functions for Next.js apps go in `/pages/api`, but all other official [runtimes](https://vercel.com/docs/concepts/functions/serverless-functions/runtimes), including Node.js and Go, use `/api`. 

You can learn more about the Go Runtime here: [vercel.com/docs/concepts/functions/serverless-functions/runtimes/go](https://vercel.com/docs/concepts/functions/serverless-functions/runtimes/go)

**Reply by wakywayne** [reply]
😄   I understand now, THANKS! I know that there are data limits on the serverless functions, but if I am on the free tier is there a limit to how many serverless functions I can have? I plan to use goLang for all my intensive data manipulation and calculations so I would like to know if I have a limit on the free tier for how many I can use.

**Reply by wakywayne** [reply]
I'm only asking because I heard someone mention the free tier has a 15-serverless function limit.

**Reply by amyegan** [reply]
Great question! For Hobby Plans, the [limit](https://vercel.com/docs/concepts/limits/overview#general-limits) is 12. But, Next has a bundling feature enabled that can help you work around the limit to some extent. 

> In the case of the "Serverless Functions Created per Deployment" limit, only dedicated Lambdas are counted as one Serverless Function. With Next.js, Vercel will automatically combine as many Serverless Functions as possible into a single Lambda, as mentioned in [bundling serverless functions](https://vercel.com/docs/concepts/functions/serverless-functions#bundling-serverless-functions).
>
> — [Limits: Serverless Functions Created per Deployment](https://vercel.com/docs/concepts/limits/overview#serverless-functions-created-per-deployment)

**Comment by coyotespike** [comment]
Sadly, I can't get this to work for love or money.

NextJS freaks out when I have only the `/api` files, or if I have `/api` and also `/pages/api`. All API routes break immediately. 

Locally, I can use Node child processes or `PythonShell` to run a Python file. On Vercel, this runs into tree-shaking issues, but by using `fs.readFileSync` I was able to force the Python files to be found. But again, when deployed, Vercel can't find Python. So running PythonShell only works locally.

Any ideas for getting this work?

**Reply by wiiiimm** [reply]
You should not use `/pages/api/` with `/api`. You will have to separate the two as separate projects (you can have them inside a monorepo but as separate projects).

**Comment by mephobic** [comment]
I have a solution as of Feb, 2023 - 

First create a Next Js project with create-next-app. Then you need to install Vercel. I've done mine with yarn.
I am working with the **app** directory of NextJs 13. 

Things to do -
1. ```api/``` folder needs to be placed in the root and ```/pages/api``` needs to be removed
2. Add requirements.txt in the root folder.
3. You need to create a new file called ```vercel.json``` and place it in the root directory along with your ```api/``` folder.
4. My **api** directory has a subfolder called v1 and within it there is **index.py**
5. Here's the vercel.json
```
{
    "routes": [
        {
          "src": "api/v1/(.*)",
          "dest": "api/v1/index.py"
        }
      ]
}
```
6. Here's my index.py file (I am using Flask as a dependency)
```from flask import Flask

app = Flask(__name__)


@app.route('/api/v1')
def home():
    return 'API Home page'


@app.route('/api/v1/about')
def about():
    return 'API About page'
```
7. **Important** - Finally run ```yarn vercel dev``` 

You should be able to access your APIs. Since both your NextJs app and your Python Backend are being deployed from the same domain, you won't even have to deal with CORS.

**Comment by spiceee** [comment]
I'm currently working on a project with both Node and Rust serverless functions. While it all works when putting both Node and Rust under the root `api/` folder, I'd prefer to have the Node ones under `pages/api`. At least one vendor package for Next.js (auth0) assumes the serverless functions are under `pages/api`, and uses `NextApiHandler` to define routes automatically. Those will break when the deployed app uses `api/`. Not a blocker, as one can define the handlers "manually" (I believe they "break" not because of NextApiHandler but bc of the route bracket definition). 

It's my understanding that somewhere in the deployment process, a redirect option is made, and it's now set as either `pages/api` or the `api/` when a custom runtime function configuration is found, but I can tell it's a custom mistake folks will make because Node functions are thought to be a vanilla Next.js thing.

**Reply by wiiiimm** [reply]
You should not use `/pages/api/` with `/api`. You will have to separate the two as separate projects (you can have them inside a monorepo but as separate projects).

**Comment by robynvasco** [comment]
So can I get a serverless function for my Next.js app in /pages/api, and a python runtime in  /api both get to work? If yes how?
If I try to run vercel dev only the /api python runtime works. It then can't load the /pages/api/generate.ts file

**Reply by wiiiimm** [reply]
You should not use `/pages/api/` with `/api`. You will have to separate the two as separate projects (you can have them inside a monorepo but as separate projects).
